### DISCUSSION TITLE
Experience with Training Custom Wakeword

### DISCUSSION THREAD

**Discussion by madgrizzle** [dscripka/openWakeWord#45] [General]
I can go into more detail later (been real busy with life) and answer any questions, but I manage to build a custom wakeword for my robot ("hey floyd") to work on openwakeword and the performance is better than any wakeword listener I've tried.  A few general observations:

1) I couldn't get the notebook code to run without error out of one thing or another.  Could easily be me considering I never tried to run a notebook before, but I couldn't get it to work.
2) I copied the code from the notebook and got it to run locally on my Windows 10 that has a GPU.  I had to install Python 3.10.. not the latest to get the dependencies to work out.
3) Running the code from the notebook locally worked until the part it tried to create batches of positive samples and would error out with an message regarding mismatched tensor sizes.  It worked when I used the data from the repo, so I suspected I needed more than a few hundred "hey floyd" samples that I had (I had a couple hundred synthetic samples and about 100 actual samples from my and my kids).
4) I decided to create more synthetic samples using [this](https://github.com/dscripka/synthetic_speech_dataset_generation) and it took a while to get it to run on my Windows machine because there was a dependency (espeak) that wasn't able to be located.  Fortunately, I found on their github a small change to the code that would avoid using enviornmental variables and it loaded.  However, after I generated 3000 samples with nvidia waveglow, I discovered all of them were completely silent.. no generated speech.  Could have easily been something on my setup so I gave up on my windows machine and switched to my ubuntu machine where this kind of them just seems easier.
5) I had no installation issues on the ubuntu machine (used a virtual environment) and used VITS to generate the 3000 samples since I had no GPU on the ubuntu machine
6) I took the 3000 synthetic samples and the 100 personal samples and used them with the training model.  No more erroring out because of differing tensor sizes.  However, I do get a RunTimeWarning about an invalid value encountered in cast, but it doesn't error out.
7) It generated the wakeword and I transferred it to my robot that uses OVOS and an openwakeword plugin and after working out a bad default setting in OVOS, it worked amazingly well.
8) Make sure the audio you send to openwakeword is sent in 1820 sample chunks (or a multiple of 1820), otherwise performance will suffer until @dscripka comes up with a solution for non-1820 multiple sample chunks.. if he decides its necessary (I question whether it is).

**Comment by sangheonEN** [comment]
Could you please share your training code?

https://github.com/dscripka/openWakeWord/blob/main/notebooks/automatic_model_training.ipynb

Did you do the training in the same way as this course?

Or, did you do the training in the same way as this course: https://github.com/dscripka/openWakeWord/blob/main/notebooks/training_models.ipynb?

I'm curious about the detailed training process and code.

**Comment by muhammadabidPro** [comment]
Nanowakeword 🔥
